nixuser: declarative user environments

ts468 opened the pull request:
What is Nix User Profile (NixUP)? NixUP is a declarative configuration for the user environment. It is an equivalent to the NixOS configuration method that is based around 'nixos-rebuild' and '/etc/nixos/configuration.nix'. NixUP provides a module system for configuring the user environment and is intended as a replacement for the imperative 'nix-env' commands. NixUP allows to, e.g., install packages, manage user defined services and to manage resources.

NixUP consists of a new program, called 'nixup' and a declarative configuration rooted at '$XDG_CONFIG_HOME/nixup/profile.nix' (e.g. ~/.config/nixup/profile.nix). Basically, the workflow for managing the NixUP user profile is the same as how the NixOS system configuration is being managed. The 'profile.nix' is edited by the user, and then turned into an active environment through 'nixup'.

The important commands for using 'nixup' are:

nixup build
  -- Builds a user profile. By default the profile is defined
     in $XDG_CONFIG_HOME/nixup/profile.nix.

nixup login
  -- Builds a user profile that will be activated on the next login.
     That is similar to nixos-rebuild boot.

nixup switch
  -- Builds a user profile and immediately switches to it.

nixup edit
  -- Opens an editor with the current configuration.

NixUP also brings an small program that helps to install and to remove software packages. If 'config.imperativeNix.enable=true' is set in the 'profile.nix' configuration, then a program 'nix-package' becomes available that manages a list of packages to be installed into the user environment. By default the list is maintained at '$XDG_CONFIG_HOME/nixup/packages.nix', from where the list is read by a module of the NixUP system.

The commands for using 'nix-package' are:

nix-package install hello

nix-package remove hello

Note that the packages are not pinned at a particular version but are linked to the currently active nixpkgs channel. A way to pin a package will be provided later.

Besides managing software packages, NixUP also provides a way to manage user controlled services. NixUP allows to define services for 'systemd --user', similarly to how NixOS allows to define services for 'systemd'. That means that the feature of managing services within NixUP is bound very tightly to 'systemd', and will not be available on other platforms like OSX or Windows.

The last interesting feature of NixUP is that is allows to manage resources. That means, it allows to define files in 'profile.nix' that are then linked into the $HOME directory of a user, or into any sub-directory of $HOME. The necessary sub-directories and links are created as needed, and automatically removed when the 'profile.nix' changes. Automatically created sub-directories are removed if they are empty after all links have been removed. So NixUP has a built-in way for managing, e.g., dot-files.

ts468 commented:
Could I get some feedback and ideas about how to name the configuration tree for nixuser, please?

@nbp Can it be merged with #4594?

geerds commented:
Great work! :+1:

jagajaga commented:
Why do we need this when we have `.nixpkgs/config.nix` where we can write our environments?
I think that's an redundant work and can confuse people.

ts468 commented:
Just to preempt a discussion about WHY we need it, don't be tricked by the simplicity of the settings that are available so far. Nixuser is intended as a central place where user environment related configuration options can be collected. So it goes beyond an individually created `.nixpkgs/config.nix`.

jagajaga commented:
That's an extra thing. I really don't see anything we don't have now that you want there.
The only thing we miss --- managing user dotfiles.

dezgeg commented:
For instance, per-user declarative systemd services (with e.g. the same start/stop and restart-if-config-changed logic that `nixos-rebuild` has) is something that needs to be in `nixuser-rebuild`.

ts468 commented:
@jagajaga maybe you want to write a nixuser module to manage dot-files?---It would be great contribution!

dezgeg commented:
The Obligatory Bikeshed: The filename `configuration.nix` is sort-of already taken, in particular `man configuration.nix` already brings up the NixOS manual. Maybe something like `~/.nixuser/user-config.nix`?

nbp commented:
> Could I get some feedback and ideas about how to name the configuration tree for nixuser, please?

I will look deeply into it later.

> @nbp Can it be merged with #4594?

I think so.

The problem I wanted to avoid is that we have to edit a configuration file to add packages.

I think `nix-env` is a convenient tool for installing new packages.  But `nixuser-rebuild` only provide a NixOS like approach, which is not a user friendly CLI for installing/uninstalling packages.

ts468 commented:
@dezgeg Thanks for pointing that conflict out! I hoped we could mirror `nixos-rebuild` as much as possible and that motivated `~/.nixuser/configure.nix`. Also, what do you think about nixup (nix user profile) instead of nixuser? It's used in #4594 and would also be a good choice. I'm just trying to collect opinions :)

@nbp Thanks for going to have a look! Let me know once I can go ahead with merging our two PRs. However, having an explicit configuration file where all installed packages have to be listed is exactly what I want to achieve! To me it is one of the most important features of NixOS that the whole configuration is in one file or place, and if possible I would like to have (exactly) the same for nixuser/nixup. And, text editors are still a very good CLI based configuration tool, i think ;) But of course you're right that a nice GUI would be highly appreciated! Maybe we can generate one semi-automatically from the module definitions, I think @matejc has been working on something like that.

jagajaga commented:
cc @edolstra

copumpkin commented:
Does this depend on NixOS, or could it work on other Nix platforms?

ts468 commented:
@copumpkin I think it should work on other platforms as well, but I haven't tested it.

nbp commented:
> @nbp Can it be merged with #4594?

Apart from the options for managing the nixpkgs configuration, and the user packages / path, it seems that our implementations can safely be merged, except for renaming integration.

Honestly, I prefer the approach taken in NixOS, which is to have a central list of build targets stored in `user.build`, and then the top-level script `nixuser-rebuild` should pull any installable targets from this attribute set.

The interesting part of NixUP is the implementation of the activation script and the resources files, which ideally should be executed by the shell at the start-up of the shell.  This activation script serves as an init-system for the user home directory.

I am not convinced by the `-rebuild` in the `nixuser-rebuild` name, as I consider this to be an implementation detail.

> However, having an explicit configuration file where all installed packages have to be listed is exactly what I want to achieve! To me it is one of the most important features of NixOS that the whole configuration is in one file or place, and if possible I would like to have (exactly) the same for nixuser/nixup. And, text editors are still a very good CLI based configuration tool, i think ;) But of course you're right that a nice GUI would be highly appreciated! Maybe we can generate one semi-automatically from the module definitions, I think @matejc has been working on something like that.

I agree, being able to define packages declaratively is nice, but the imperative style provided by `nix-env` is convenient.

What I was thinking about was to let the CLI tool generate the modules which are adding the packages in the environment.

ts468 commented:
@nbp Thanks for your feedback, I'll take it into account! I already had a closer look at how to merge the two implementations, and I have to say, I just started to fully understand and appreciate all the nice details that you put into NixUP! I think you've done an amazing job! Thank you very much for putting it together, it would be a shame if your work wouldn't make it upstream. I just hope I don't break too much while merging the two PRs.

I was reading up on the mailing list again about your introduction of your PR, especially the question when to activate a new user environment. I think it should be possible to have different options like `nixup switch` or `nixup login` or `nixup shell` to distinguish between an immediate activation, an activation with a new login or an activation just for the current shell. What do you think?

I like the idea about your CLI very much. What about a hidden configuration file _outside_ of the nixup/nixuser configuration tree that is only managed by an CLI, and then sourced appropriately? I could imagine to have, e.g., ~/.nixup/profile/managed/{permanent,temporary} where packages are added or removed. The difference could be that the temporary packages are only available until the next rebuild of the declarative environment, whereas permanent packages are kept until they are removed. Could that work?

nbp commented:
> I was reading up on the mailing list again about your introduction of your PR, especially the question when to activate a new user environment. I think it should be possible to have different options like `nixup switch` or `nixup login` or `nixup shell` to distinguish between an immediate activation, an activation with a new login or an activation just for the current shell. What do you think?

The problem, is that as opposed to `NixOS`, we can have multiple instances of the shell, which can share some common configuration, such as agent program which are running in the background, as long as there is one session open.

Doing it for the current shell sounds interesting, but then the question is how do you ensure that the configuration files are only visible within the current shell and only within any application started by this shell.

Since I read about `buildFHSUserEnv`, I was wondering if we could use a mount-point namespace solution to make such testing mechanism possible.  Then the question, is why do we have a different solution for the testing mechanism and the standard mechanism.  If we are going that way, then we should do it for everything.  The problem, is that mount-point namespaces are a Linux only feature, as far as know, which makes this unpractical for MacOS users.

> I like the idea about your CLI very much. What about a hidden configuration file outside of the nixup/nixuser configuration tree that is only managed by an CLI, and then sourced appropriately?

Yes, that's what I was thinking about, which is to have a `~/.nixup/generated/user.packages.firefox.nix` module generated when you install firefox from the command line.  And do the same for every option which is defined from the command line.

Basically, the name of the file is used as a simple way to manipulate each definition without having any knowledge on how modules are written & handled.

> I could imagine to have, e.g., `~/.nixup/profile/managed/{permanent,temporary}` where packages are added or removed. The difference could be that the temporary packages are only available until the next rebuild of the declarative environment, whereas permanent packages are kept until they are removed. Could that work?

That's sounds doable, but I would not worry about that right now.

ts468 commented:
I've pushed an updated version of the declarative user environment. I tried to merge with #4594. Have a look and give me some feedback, please!

bjornfor commented:
(I haven't had the time read everything, so I hope I'm not saying something silly, or something already suggested.)

Is it planned to be able to hook into user environments from /etc/nixos/configuration.nix?

I'm thinking that if "~/.nixup/profile.nix" contains

```
{ config, pkgs, ... }:
{
  user.packages = with pkgs; [ vim firefox thunderbird ];
}
```

then a similar option can exist in /etc/nixos/configuration.nix (options gets merged):

```
{ config, pkgs, ... }:
{
  users.extraUsers.<name>.packages = with pkgs; [ vim firefox thunderbird ];
}
```

Basically, options in "~/.nixup/profile.nix" can be rooted in configuration.nix at "users.extraUsers.\<name\>.*" (or something like it).

ts468 commented:
@bjornfor Thanks for your interest! Indeed, it is planed to make all NixUP options that are available under "~/.nixup/profile.nix" also available in the NixOS configuration, probably under "users.extraUsers.<name>.profile". However, I'm just not so sure how to merge the options defined in NixOS and NixUP, yet. Should they just be added up? Should the options defined in NixOS be defaults that can be overwritten with NixUP, or should that be prohibited? If you have any suggestions or ideas that would be great! :)

nbp commented:
> Basically, options in "~/.nixup/profile.nix" can be rooted in configuration.nix at "users.extraUsers.<name>.*" (or something like it).

This can be done quite easily, as by design modules are valid submodules.

benley commented:
I would love to be able to use something like this on non-NixOS systems, fwiw.  One can approximate it with buildEnv in ~/.nixpkgs/config.nix, but it would be very, very nice to have something that actually handles user dotfiles.

anderspapitto commented on nixup/lib/eval-config.nix:
```diff
@@ -0,0 +1,58 @@
+# as subcomponents (e.g. the container feature, or nixops if network
+# expressions are ever made modular at the top level) can just use
+# types.submodule instead of using eval-config.nix
+{ # !!! system can be set modularly, would be nice to remove
```
can these first three (system, pkgs, baseModules) already be addressed (which I guess would mean removed, based on the comments?)

anderspapitto commented on nixup/lib/from-env.nix:
```diff
@@ -0,0 +1,4 @@
+# TODO: remove this file. There is lib.maybeEnv now
```
can this already be done?

anderspapitto commented on nixup/modules/config/nix.nix:
```diff
@@ -0,0 +1,20 @@
+
+    nix.package = mkOption {
+      type = types.package;
+      default = pkgs.nix;
```
what's another possible value that someone might want to set this to?

joachifm replied:
`pkgs.nixUnstable`

anderspapitto commented on nixup/modules/config/packages.nix:
```diff
@@ -0,0 +1,75 @@
+      '';
+    };
+
+    user.pathsToLink =
```
is this dead code due to the commented-out line 10 above?

anderspapitto commented on nixup/modules/config/resourceFiles.nix:
```diff
@@ -0,0 +1,90 @@
+      default = {};
+      type = types.loaOf (types.submodule options);
+      example = literalExample ''
+        [ { target = ".ssh/config";
```
could this be simplified to an attribute set where the key is the file name and the value is the string contents? Nix already has utilities for e.g. reading in file contents as a string (bulitins.readFile), so it seems unecessary to reinvent that with the "source" option

anderspapitto commented on nixup/modules/misc/legacynixpkgsconfig.nix:
```diff
@@ -0,0 +1,61 @@
+
+    in
+      # allow both:
+      # { /* the config */ } and
```
it would probably save some copy-pasting beginners some headache to only have one way (the second one)

anderspapitto commented on nixup/modules/nixup/nixup.sh:
```diff
@@ -0,0 +1,203 @@
+# If the Nix daemon is running, then use it.  This allows us to use
+# the latest Nix from Nixpkgs (below) for expression evaluation, while
+# still using the old Nix (via the daemon) for actual store access.
+if systemctl show nix-daemon.socket nix-daemon.service | grep -q ActiveState=active; then
```
systemctl is-active nix-daemon.socket
systemctl is-active nix-daemon.service

anderspapitto commented on nixup/modules/nixup/nixup.sh:
```diff
@@ -0,0 +1,203 @@
+fi
+
+
+# First build Nix, since NixOS may require a newer version than the
```
Can this logic be shared with the same logic for the NixOS-level?

anderspapitto commented on nixup/modules/nixup/switch-to-configuration.pl:
```diff
@@ -0,0 +1,458 @@
+            # Stopping a target generally has no effect on other units
+            # (unless there is a PartOf dependency), so this is just a
+            # bookkeeping thing to get systemd to do the right thing.
+            if (boolIsTrue($unitInfo->{'X-StopOnReconfiguration'} // "no")) {
```
Is all of this systemd-related logic copy-pasted from NixOS-level code?

jagajaga commented on nixup/modules/nixup/switch-to-configuration.pl:
```diff
@@ -0,0 +1,458 @@
+#! @perl@
```
We are trying to get rid of perl in nix...

geerds replied:
@jagajaga What's the preferred language? Python?

jagajaga replied:
C++ 
See this: https://www.bountysource.com/issues/4399291-get-rid-of-the-perl-dependency and https://github.com/NixOS/nix/issues/341

dezgeg replied:
Isn't that just for the nix core, not NixOS/nixpkgs

jagajaga replied:
I think this issue can be related to this module.

edolstra replied:
@jagajaga We may be trying to get rid of Perl in Nix, but not in NixOS.

nbp commented:
The first version does not have to be perfect, it just has to handle basic features on which we can iterate.

If you want to eliminate dead code, rewrite everything in another language, share the logic with other part of the repository, … then we can still do these after these modifications are merged.

The most important part is to have something which is working, and merged.
Do anybody see any blocker which might prevent this from working?

dezgeg commented on nixos/modules/config/nixup.nix:
```diff
@@ -0,0 +1,66 @@
+
+        # Activate nixup user profile.
+        if test -e $NIXUP_USER_PROFILE_DIR/default/activate; then
+            HOME=$PAM_HOME $NIXUP_USER_PROFILE_DIR/default/activate
```
Isn't this executing a user-controlled script as root?

dezgeg commented on nixos/modules/installer/tools/nixos-install.sh:
```diff
@@ -219,6 +219,7 @@ NIX_PATH="nixpkgs=/tmp/root/$nixpkgs:nixos-config=$NIXOS_CONFIG" NIXOS_CONFIG= \
 # Copy the NixOS/Nixpkgs sources to the target as the initial contents
 # of the NixOS channel.
 mkdir -m 0755 -p $mountPoint/nix/var/nix/profiles
+mkdir -m 1777 -p $mountPoint/nix/var/nix/profiles/nixup
```
Do we actually need a new subdirectory for nixup profiles? For instance, per-user channels use e.g. `/nix/var/nix/profiles/per-user/root/channels`

wmertens replied:
👍 for using per-user/$USER/nixup. It won't conflict with what's there and it is backwards compatible with current installations.

ts468 commented:
Thank you all very much for your feedback! Especially @nbp and @anderspapitto for your detailed review, it's highly appreciated! I will get back to it as soon as possible.

As Nicolas said, the current aim is to get something merged that we can build up on. That means, I'm particularly interested in suggestions concerning the "front-end", that is the interface that nixup modules are written against. The "back-end" stuff is mostly copied from NixOS and can be changed later without too much disruptions, I think.

Concerning the question of how to support non-NixOS systems. I think that basic functions like, e.g., the abstractions around nixpkgs, might be quite NixOS-independent, and so could work on other systems. When it comes to the support of nixup user services, I'm not so sure though. Like NixOS, it will be based around systemd, like it or not. So in order to make that feature available on non-NixOS systems, there would have to be a way to run a `systemd user session`, or something compatible at least. I don't know if projects like `systembsd` could be useful for that. But that won't be anything that I'll be working on.

dezgeg commented on nixup/modules/nixup/nixup.sh:
```diff
@@ -0,0 +1,203 @@
+}
+
+
+# Add a sensible default value for <nixup-config> to NIX_PATH.
```
`nix-instantiate --find-file nixup-config` could help to do this properly.

jagajaga commented:
I insist on rewriting to c++ now, before merging. Because if we merge now that will not soon be rewritten.

ts468 commented:
@jagajaga I think it will be quick to switch the "back-end" of NixUP to C++ once it is available from NixOS. If you insist on C++ right away, how long will it take YOU to provide a working solution?

edolstra commented:
Hm, does this actually need to be merged into the Nixpkgs repo? Any reason why it can't work fine in  a branch or a separate repository? I ask because I'm hesitant about merging these kinds of major experimental additions into the Nixpkgs repo, until we're confident that a) we're convinced this is the right approach; and b) we want to commit to maintaining it. (Plenty of examples in the Nixpkgs repo of stuff that ended up bitrotting or only used by one person...)

Technical comment: there is a lot of cut&paste in this PR, in particular `switch-to-configuration.pl` and the systemd modules. That's obviously pretty bad for maintainability.

Also, the PR seems to contain a lot of stuff that has nothing to do with the PR (such a openxpki and the Perl modules).

anderspapitto commented:
if/when this is merged into NixOS, all of the systemd.services.user.* options should probably be removed/deprecated, so that user-level systemd services are configured through a single mechanism

anderspapitto commented:
Also

```
[anders@gurney:~]
$ wget -q https://raw.githubusercontent.com/ts468/nixpkgs/upstream.nixuser/nixup/modules/nixup/switch-to-configuration.pl
[anders@gurney:~]
$ md5sum switch-to-configuration.pl 
5554ec89848531c8bcb1cc66b4a9607d  switch-to-configuration.pl
[anders@gurney:~]
$ md5sum /etc/nixos/nixpkgs/nixos/modules/system/activation/switch-to-configuration.pl 
5554ec89848531c8bcb1cc66b4a9607d  /etc/nixos/nixpkgs/nixos/modules/system/activation/switch-to-configuration.pl
```

there is certainly some low-hanging fruit with regards to code reuse

edwtjo commented:
> if/when this is merged into NixOS, all of the systemd.services.user.* options should probably be removed/deprecated

This is highly debatable. It is perfectly valid for an admin to want to provide predefined user services for its users through `configuration.nix`.

copumpkin commented:
Any progress on this?

ts468 commented:
I hope to have some discussions about it during the NixOS Conference. Maybe I can make some progress during the sprint after the conference.

peti commented:
I'd prefer the configuration files to live in `~/.nixpkgs` by default. There's too many bloody dot.something files/dirs already, and I don't think there's a compelling technical reason to create yet another one for this application when Nix already introduced such a name.

edolstra commented:
Let's make that ~/.config/nixpkgs to take it into the 21st century :-)

geerds commented:
Yes, `$XDG_CONFIG_HOME` would be a good idea!

wmertens commented:
I would prefer the directory to be named `.nix`, and on OS X it would be best to put it at `~/Library/Nix`.

ts468 commented:
An updated version of NixUP ready to be tested :)

That is included:
 - [✓] Generate an activation script.
 - [✓] Abstract the creation of resource files.
 - [✓] Option to list installed packages.
 - [✓] Command line tool to manage profiles generations and updates.
 - [✓] Command line tool to install / remove packages without using an editor. 

Please have a look at NIXUP_README.md for the commands to set up and run a NixOS container with NixUP included.

Please give me some feedback!

globin commented:
I'd agree with @geerds and @edolstra on ~/.config/nix[whatever]

ts468 commented:
@edolstra How should we move NixUP forward? I guess you would have to review the integration into NixOS, or? :) Is there anything I can do to help reviewing the code?
Maybe for now you would only need to review the changes that are made within the nixpkgs/nixos directory. There is some code that I copied from nixpkgs/nixos into nixpkgs/nixup to minimize the initial changes in nixpkgs/nixos. I would try to remove the duplicated code once NixUP is merged, so you wouldn't have to review all the parts under nixpkgs/nixup in too much detail. But of course I'm happy to get as much feedback as possible :)

edolstra commented on nixos/modules/security/pam.nix:
```diff
@@ -93,6 +94,19 @@ let
         '';
       };
 
+      pamExec = mkOption {
```
`pamExec` -> `sessionCommands` or something like that. `pam` is superfluous because we're already in the PAM options, and `exec` is vague.

edolstra commented on nixos/modules/security/pam.nix:
```diff
@@ -264,7 +278,7 @@ let
 
           # Session management.
-          session required pam_env.so envfile=${config.system.build.pamEnvironment}
+          session required pam_env.so readenv=0 conffile=${config.system.build.pamEnvironment}
```
Why the change from `envfile` to `conffile`?

edolstra commented on nixos/modules/programs/shell.nix:
```diff
@@ -37,6 +38,28 @@ in
+              echo "  };"                             >> $XDG_CONFIG_HOME/nixup/profile.nix
+              echo ""                                 >> $XDG_CONFIG_HOME/nixup/profile.nix
+              echo "}"                                >> $XDG_CONFIG_HOME/nixup/profile.nix
+          fi
```
Maybe something like
```
let defaultProfile = pkgs.writeText ...; in
''
  cat ${defaultProfile} > $XDG_CONFIG_HOME/nixup/profile.nix
''
```
would be nicer.

edolstra commented on nixos/modules/config/nixup.nix:
```diff
@@ -0,0 +1,75 @@
+      type = types.bool;
+      default = true;
+      description = ''
+        Whether to integrate NixUP into NixOS or not. If
```
=> `Whether to enable NixUP`.

jgeerds commented on nixos/modules/config/nixup.nix:
```diff
@@ -0,0 +1,75 @@
+
+  options = {
+
+    nixup.enable = mkOption {
```
You could also use `mkEnableOption` (see **lib.options** for more information)

edolstra commented on nixos/modules/programs/environment.nix:
```diff
@@ -21,6 +21,8 @@ in
         PAGER = mkDefault "less -R";
         EDITOR = mkDefault "nano";
+        XDG_CONFIG_HOME = "$HOME/.config";
+        XDG_DATA_HOME = "$HOME/.local/share";
```
Since these are the default values according to the XDG spec, should we really set them explicitly? Might be nicer not to clutter the environment.

bennofs replied:
AFAIK, there's some broken software that won't work correctly without explicitly setting them. I don't remember exactly which applications, but I set them explicitly in my config.

edolstra commented on nixos/modules/programs/environment.nix:
```diff
@@ -8,7 +8,7 @@ with lib;
 let
 
-  cfg = config.environment;
+  cfg = config;
```
Better to remove `cfg` entirely.

edolstra commented on nixup/modules/config/packages.nix:
```diff
@@ -0,0 +1,80 @@
+  config = {
+    nixup.build.userPath = userPath;
+    nixup.buildCommands = ''
+      echo "DONE" >> $out/info
```
Remove?

edolstra commented on nixup/modules/module-list.nix:
```diff
@@ -0,0 +1,13 @@
+  ./nixup/activation.nix
+  ./nixup/build.nix
+  ./nixup/imperativenix.nix
+  ./nixup/resourceFiles.nix
```
No camelCase in file names please.

edolstra commented on nixup/modules/module-list.nix:
```diff
@@ -0,0 +1,13 @@
+  ./config/nix.nix
+  ./config/packages.nix
+  ./misc/assertions.nix
+  ./misc/legacynixpkgsconfig.nix
```
=> `legacy-nixpkgs-config.nix`.

edolstra commented on profile.nix:
```diff
@@ -0,0 +1,41 @@
+{ config, pkgs, ... }:
```
This looks like an example file that should be removed or moved to the documentation.

edolstra commented on pkgs/top-level/all-packages.nix:
```diff
@@ -219,6 +219,8 @@ let
 
   nixpkgs-lint = callPackage ../../maintainers/scripts/nixpkgs-lint.nix { };
 
+  nixup = (callPackage ../../nixup/modules/nixup/tools.nix { }).config.nixup.build.nixup;
```
Will this cause an entire module system evaluation when running `nix-env -qa`?

edolstra commented on nixup/modules/nixup/nixup-resources.sh:
```diff
@@ -0,0 +1,124 @@
+	DIR=$2
+	if [ "$DIR" != "." -a -d "$BASE/$DIR" ]; then
+		if [ ! "$(@coreutils@/bin/ls -A $BASE/$DIR)" ]; then
+			@coreutils@/bin/rm -rf $BASE/$DIR
```
`rm -rf` on anything in my home directory frightens me. Especially when the arguments aren't quoted.

bennofs replied:
Also, isn't this a possible race condition, if a file is created between the ls call and the rm call?

wmertens replied:
This can instead be done with `rmdir` without the `if`, since that will error out if the dir is missing or not empty.

edolstra commented:
Some comments:

* Yeah, please unify the duplicated files (especially the systemd stuff).

* This PR introduces several NixUP-specific files in the root directory, such as `Makefile`, `NIXUP_README.md`, `configuration_nixup.nix` and `profile.nix`. These should be moved into `nixup/`.

* Using a Makefile for purely imperative actions seems weird to me. IMHO it's best to get rid of that Makefile. If it's just to help people set up a container for testing, it's better to be explicit in the docs (e.g. just tell people to run `nixos-container start nixup` etc.).

* It might be possible to unify `nixup-resources.sh` with `nixos/modules/system/etc/setup-etc.pl` since they appear to serve a similar purpose. The latter has the advantage of being written in a sane language where you don't have to worry about what `rm -rf $BLA` will do when $BLA contains whitespace, etc. It already looks like `user.resourceFiles` has the same type as `environment.etc`.

* Maybe rename `user.resourceFiles` to `user.files` or something like that? 

* I don't get the name `imperativeNix`, since it apparently manages a declarative configuration. `nix-env -i` == imperative, `user.packages` == declarative, and `imperativeNix` is just a tool for editing the latter.

* `nix-package install` => `nixup add-package` or something like that, etc.

ts468 commented:
@edolstra Thank you so much for your time and efforts for reviewing NixUP! :) Your comments and suggestions are very helpful! I'll get back to you once I've worked through your list ;) Thanks!

Profpatsch commented:
Are there any news on this?

ts468 commented:
It has been reviewed, and I know what to do. Hopefully I can merge NixUP for 16.03.

gleber commented:
I love this! That will be the best ever user environment manager ever. I'm very eager to integrate it into my dotfiles.

jgeerds commented:
@ts468 any progress on this? :smile:

zimbatm commented:
@ts468 I implemented a couple of the review suggestions above, feel free to cherry-pick whatever you like: https://github.com/zimbatm/nixpkgs/tree/upstream.nixuser

layus commented:
A nice addition to user services would be to allow lingering. Loginctl supports this, but it is not yet available in nixos config. See #3702.

therealpxc commented:
Add me to the list of those interested in seeing this merged ASAP. This looks great! NixUP is a killer feature for me.

Azulinho commented:
+1

ericbmerritt commented:
+1

gilligan commented:
*bump* ;) 

What is the status on this? I just asked about something similar on irc and this PR came up.

gilligan commented:
@zimbatm ? @ts468 ?

ts468 commented:
Thank you all for your interest and help on nixup! I'm very sorry that I couldn't work actively on it for so long, I feel really bad about that. I can't tell you exactly when, but I will definitely get back to it, probably in June.

zimbatm commented:
I think for this to happen it needs to be broken down into smaller steps that can be individually merged. It's also not entirely clear if this is the right approach because I think it touches the user's home which is a bit scary in itself.

gilligan commented:
Maybe we should address it on the mailing list to get input.

Azulinho commented:
" I really want this"

On 3 May 2016 at 20:57, Tobias Pflug <notifications@github.com> wrote:

> Maybe we should address it on the mailing list to get input.
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <https://github.com/NixOS/nixpkgs/pull/9250#issuecomment-216646542>
>

sorpaas commented:
+1

therealpxc commented:
I'm in the process of Nixifying my home network right now and one of the things I want to do is start using nixup (based on the revisions by @zimbatm to the work @ts468 has done) to manage some local configuration stuff.

Currently, I use a tool called [super-user-spark](https://github.com/NorfairKing/super-user-spark) (also [in nixpkgs](https://github.com/NixOS/nixpkgs/blob/7475728593a49f09c4b7b959b15513aee38ab4b4/pkgs/applications/misc/super_user_spark/default.nix)) to manage my dotfiles. It has a lot in common with other simple link-based tools for dotfile management, but what I love about it is its [very simple syntax](https://github.com/NorfairKing/super-user-spark/blob/master/doc/language.md). By default, it won't overwrite files that already exist, and it can use hardlinks as well as symlinks.

Anyway, I'm going to try switching some of my dotfile management over to nixup. I'll let y'all know how it goes, and of any changes I end up making along the way.

Profpatsch commented:
> Currently, I use a tool called super-user-spark (also in nixpkgs) to manage my dotfiles. It has a lot in common with other simple link-based tools for dotfile management, but what I love about it is its very simple syntax. By default, it won't overwrite files that already exist, and it can use hardlinks as well as symlinks.
> One thing I'm thinking about is maybe having nixup use SUS

There is an “official” GNU tool that does something similar, it’s called [stow](https://www.gnu.org/software/stow/stow.html).

Azulinho commented:
I am using nix to deploy services on Linux distros that don't provide
packages for those services.

example below is for apache mesos through nix, I deploy this to a suse12
box.
nix packages are installed by the user nix and I then add systemd startup
services to execute them.

https://github.com/Azulinho/ci_on_mesos_on_centos7/blob/ansibling_it/roles/nix-mesos-master/tasks/main.yaml

it would be great to be able to have the ability to run those services
directly through nix.
On 7 May 2016 11:50, "Profpatsch" <notifications@github.com> wrote:

Currently, I use a tool called super-user-spark (also in nixpkgs) to manage
my dotfiles. It has a lot in common with other simple link-based tools for
dotfile management, but what I love about it is its very simple syntax. By
default, it won't overwrite files that already exist, and it can use
hardlinks as well as symlinks.
One thing I'm thinking about is maybe having nixup use SUS

There is an “official” GNU tool that does something similar, it’s called
stow <https://www.gnu.org/software/stow/stow.html>.

—
You are receiving this because you commented.
Reply to this email directly or view it on GitHub
<https://github.com/NixOS/nixpkgs/pull/9250#issuecomment-217628399>

wmertens commented on nixos/modules/programs/environment.nix:
```diff
@@ -55,14 +57,15 @@ in
          export NIX_USER_PROFILE_DIR="/nix/var/nix/profiles/per-user/$USER"
-         export NIX_PROFILES="${concatStringsSep " " (reverseList cfg.profiles)}"
+         export NIX_PROFILES="${concatStringsSep " " (reverseList cfg.environment.profiles)}"
+         ${optionalString cfg.nixup.enable "export NIXUP_CONFIG=\"$XDG_CONFIG_HOME/nixup/profile.nix\""}
```
Likewise, this can simply be the default so we don't clutter the user environment

wmertens commented on nixos/modules/security/pam.nix:
```diff
@@ -313,6 +329,12 @@ let
+    #!${pkgs.stdenv.shell} -eux
+
+    ${content}
+    '';
```
nitpick: align the closing tag with the opening line indentation

wmertens commented on nixos/modules/services/x11/display-managers/gdm.nix:
```diff
@@ -133,15 +133,15 @@ in
 
         session  required       pam_succeed_if.so audit quiet_success user = gdm
-        session  required       pam_env.so envfile=${config.system.build.pamEnvironment}
+        session  required       pam_env.so readenv=0 conffile=${config.system.build.pamEnvironment}
```
Since the default of readenv is 1, perhaps people are relying on .pam_environment being read by PAM. Is there a reason these changes are part of this PR?

wmertens commented on nixos/modules/virtualisation/container-config.nix:
```diff
@@ -11,7 +11,7 @@ with lib;
     powerManagement.enable = mkDefault false;
 
-    networking.useHostResolvConf = true;
+    networking.useHostResolvConf = mkDefault true;
```
While I agree this is a useful change, I don't see how this relates to this PR.

anderspapitto commented:
something that wasn't obvious to me before trying this out, is that the nixup-managed profile destructively replaces the default user profile - so any packages that have been installed through `nix-env -i` will vanish

Azulinho commented:
as a user I would be happy with that.
my only use case of nix-env is that I can't install nix packages as a normal user without it, however the imperative form of nix-env lacks all the perks of using a declarative and functional pkg manager like nix.
I would happily give nix-env away if in return I got a way to manage my packages AND my conf files as a user through a .nix configuration file.

kamilchm commented:
You can look at my comment from https://github.com/NixOS/nixpkgs/issues/1750#issuecomment-186385710 if you need something now without NixUP.

anderspapitto commented:
@ts468 I made a pr against your branch that fixes the lowest-hanging fruit for code duplication by symlinking a few files that are bit-for-bit copies

https://github.com/ts468/nixpkgs/pull/1

the three new files that I observed which are very similar but not identical to existing files are

eval-config.nix
nixpkgs.nix
systemd-unit-options.nix

but I didn't touch those here for simplicity

anderspapitto commented:
another non-obvious point is that nixup operations aren't atomic in the way that nixos operations are, because you have to modify multiple files in $HOME - you can't do it as switching a single root symlink, unless maybe if nixup took ownership of the whole $HOME directory

nathan7 commented:
@anderspapitto /etc is similar, though! We could have an equivalent to /etc/.static, perhaps?

therealpxc commented:
What if we let the dotfile management be handled in the following way:

Each generation of a Nix home directory consists of a union filesystem where the lower layer is generated entirely by Nix, and the upper layer contains user changes. These directories, union mounted, each get a store path. If there are any conflicts between the current upperdir and the next-generation lowerdir, Nix spits out an error and doesn't switch your home directory over to the new generation. Then we can either have home directories ordinarily be empty, and bind mount whole home directories (the union mounts) from the Nix store, or convert the home directory itself into a symlink, and for new generations of the home directory, point the link to the new fused home directory in the Nix store?

akavel commented:
@therealpxc Does Nix support union filesystems? And aren't Nix store paths mounted readonly?

therealpxc commented:
@akavel that's true. Only the lowerdir is generated by Nix, so maybe that's the only piece that goes in the Nix store.

OverlayFS is built into the Linux kernel. Does OS X have any overlay/union filesystems?

akavel commented:
@therealpxc Interesting, didn't know about OverlayFS, thanks!

Azulinho commented:
@kamilchm regarding  https://github.com/NixOS/nixpkgs/issues/1750#issuecomment-186385710

I already use collections to manage my user packages and some of my config files, works great, however collections still fail to address my usage case of having services defined within my user environment.
see https://github.com/NixOS/nixpkgs/pull/9250#issuecomment-217630034

anderspapitto commented:
so, one actual bug/issue with this is that when defining user units, there's no equivalent to`source ${config.system.build.setEnvironment}`, without which a bunch of stuff doesn't work, like connecting to dbus. See for example various people's configs linked from https://github.com/NixOS/nixpkgs/issues/15807

rasendubi commented on nixup/modules/config/packages.nix:
```diff
@@ -0,0 +1,80 @@
+      description = ''
+        The set of packages that appear in
+        <filename>$XDG_RUNTIME_DIR/nixup/active-profile/sw</filename>. These packages are
+        are automatically updated every time you rebuild the user profile.
```
`are are`

rasendubi commented on nixup/modules/module-list.nix:
```diff
@@ -0,0 +1,13 @@
+  ./misc/legacynixpkgsconfig.nix
+  ./nixup/activation.nix
+  ./nixup/build.nix
+  ./nixup/imperativenix.nix
```
=> `imperative-nix.nix`.

siddharthist commented:
@ts468 Seems like there's a lot of small/misc comments on grammar, spelling, etc. Do you need help fixing those up? I'm happy to branch/PR.

sorpaas commented:
Is there any update for NixUP?

mrobbetts commented:
Definitely looking forward to this!

gpyh commented:
I'll join the chorus of "This is awesome" comments.
I would also love some status update.

zimbatm commented:
Have a look at https://github.com/sheenobu/nix-home, it seems to be maintained

jbaum98 commented:
I'm a little confused about the features this pull request provides. I thought that it brought to the user level the declarative API that was formerly only available on the system level, so that packages can be specified in a config file and nix will download them, build them, and make them all available. In that sense it replaces the imperative `nix-env`. Is that all correct?

But it also seems like this pull request adds utilities for managing dotfiles; why does that go with the feature I described above? Could they be two pull requests, or are the fundamentally related somehow? Are there other features other than that "main" one, and how are they related two that main feature such that they couldn't be split into separate pull requests?

CMCDragonkai commented:
@jbaum98 It's more to do with bringing user-profile configuration inline with NixOS system configuration. Yes there's already functionality in NixOS to install user-packages based on `~/.nixpkgs/config.nix`, but there's no functionality to manage `.dotfiles`.

ashyon commented:
How can we make this happen? Is there still hope that this PR will ever get merged?

erlandsona commented:
Any updates?

0xABAB commented:
@ts468 Lost interest?

gleber commented:
This is relevant: https://github.com/rycee/home-manager

ts468 commented:
Hi everyone!

It's been quite a while since NixUP got proposed as a declarative configuration for user environments, and despite my constant lack of time to push it forward, I always hoped I could get back to working on it at some point.

Now it seems as I could manage to re-start working on NixUP, but before doing so I wanted to ask who would still be interested in it? In particular, I would be curious to know who would like to see progress of NixUP to the point to be willing to donate 5-10 Euro to the NixOS Foundation? Don't get me wrong, I don't want to be paid for working on it, but for making NixUP a success, it will need quite some effort from quite a few rather busy people, and so I think it is just fair to ask for more than just a quick and cheap "thumbs-up".

So if you could maybe just reply with the amount you'd be willing to donate, including "0 Euro" of course, that would be great and very helpful in judging where to take NixUP.

Thanks, and very best,
Thomas

PS: I've been in contact with Robert Helgesson (rycee) who is working on the amazing [Home Manager](https://github.com/rycee/home-manager), and I hoped that the Home Manager project could benefit from some of the future work on NixUP as well. In general, we're both interested in working towards a common solution as far as possible.

therealpxc commented:
I am glad that you are interested in collaborating with Robert on
home-manager, and that you are interested in resuming this work! I would
donate 10 euros to the NixOS foundation just to see this and related work
pick up steam. If this feature makes it into a major release, I'll give
another 20 to celebrate.

On Mon, Aug 7, 2017, 11:14 PM ts468 <notifications@github.com> wrote:

> Reopened #9250 <https://github.com/NixOS/nixpkgs/pull/9250>.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/NixOS/nixpkgs/pull/9250#event-1196600081>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ADCJbmp7wEYAEJ-9HOkaGu278RYVjXnEks5sV_y8gaJpZM4FrpBB>
> .
>

casey commented:
I would happily donate 50 euro. I find the difference between package management for system and user environments pretty confusing and would be very happy if I could use declarative configurations for everything.

cumber commented:
I would happily donate 30 euro towards this

gleber commented:
50 EUR from me

On 8 Aug 2017 8:26 AM, "Casey Rodarmor" <notifications@github.com> wrote:

> I would happily donate 50 euro. I find the difference between package
> management for system and user environments pretty confusing and would be
> very happy if I could use declarative configurations for everything.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/NixOS/nixpkgs/pull/9250#issuecomment-320862629>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AACBocYwFJAgF0TGRNic7dzvBSQ3C3Slks5sV_-egaJpZM4FrpBB>
> .
>

wmertens commented:
I will match the median donation :)

On Tue, Aug 8, 2017 at 9:16 AM Gleb Peregud <notifications@github.com>
wrote:

> 50 EUR from me
>
> On 8 Aug 2017 8:26 AM, "Casey Rodarmor" <notifications@github.com> wrote:
>
> > I would happily donate 50 euro. I find the difference between package
> > management for system and user environments pretty confusing and would be
> > very happy if I could use declarative configurations for everything.
> >
> > —
> > You are receiving this because you commented.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/NixOS/nixpkgs/pull/9250#issuecomment-320862629>, or
> mute
> > the thread
> > <
> https://github.com/notifications/unsubscribe-auth/AACBocYwFJAgF0TGRNic7dzvBSQ3C3Slks5sV_-egaJpZM4FrpBB
> >
> > .
> >
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/NixOS/nixpkgs/pull/9250#issuecomment-320871704>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AADWliPGh40jbpV_r6TdN45bSvxHnTt-ks5sWAs3gaJpZM4FrpBB>
> .
>

yacinehmito commented:
20€ here.

johnramsden commented:
This is just a user speaking, but I am incredibly interested in the idea 
of user configuration. In fact it's the biggest thing I miss in nix. 
Right now I'm trying to configure everything in configuration.nix, It 
would be much cleaner to try and do it on a user level. Being able to do 
things like edit the users shellrc, ssh config, gpg config, openbox 
config, etc, would be really cool.

I'd happily donate 50€ if it helps. Do you just want people to contribute as mentioned on the [NixOS Community Page](https://nixos.org/nixos/community.html)?

kristoff3r commented:
I'd also happily donate 30€

np commented:
50€ here

On Thu, Aug 10, 2017, 23:02 Kristoffer Søholm <notifications@github.com>
wrote:

> I'd also happily donate 30€
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/NixOS/nixpkgs/pull/9250#issuecomment-321672501>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAAVrHBiOTPTRdXIGlPvxYXXWLyLmzP4ks5sW2_3gaJpZM4FrpBB>
> .
>

binarin commented:
10 euro. And another 20 when I'll be able to fully deploy my workstation using only `configuration.nix`

zraexy commented:
I will donate 85€ as this is a much needed feature.

basvandijk commented:
I would love to see NixUP happen and donated €20.

olynch commented:
I would be willing to donate 5€ if this happens.

Radvendii commented:
I'd give 5€. I'm also happy to help with the coding.

nonsequitur commented:
I've just donated 40€. Make it happen!

joshuafcole commented:
10€. Separate dotfiles is a pain that nix can address.

markus1189 commented:
Min 10€, this would be awesome!

bjornfor commented:
I've donated 50 EUR to the NixOS Foundation. (I'd like to see NixUP/Home manager integrated with Nix eventually. One CLI to rule them all.)

WhittlesJr commented:
I'd give an easy 50 bucks. Brand new to NixOS but I fell in love so quickly. As others have said, user management is the major stumbling block for me right now in my learning curve.

olejorgenb commented:
I would easily give 10€

4z3 commented:
FWIW, lots of dotfiles can be [already](https://github.com/NixOS/nixpkgs/pull/25712) replaced by wrappers like:

```nix
{
  users.users.root.packages = [
    (pkgs.writeScriptBin "htop" ''
      #! ${pkgs.bash}/bin/bash
      export HTOPRC=${pkgs.writeText "htoprc" ...}
      exec ${pkgs.htop}/bin/htop "$@"
    '')
  ];
}
```

Downside of the wrapper approach:
- There might be applications that don't provide means to specify configuration. One could override `$HOME`, but then there might be applications that require `$HOME` for other stuff than configuration. (Personally, I haven't encountered any of these, yet.)
- Applications cannot write their configuration anymore, e.g. `htop` will just terminate without error and nothing changed. (I suspect NixUP won't handle that either.)

nicknovitski commented:
I'm currently trying out an approach leveraging [rcm](https://github.com/thoughtbot/rcm).  I write a custom "dotFiles" package in my `~/.nixpkgs`.  The source includes both static files and nix code that produces dynamic ones, all of which end up in `$out` at the end.  Then I write an `install.nix` with no build steps, but a shellHook `${rcm}/bin/rcup -v -d ${dotFiles}; exit`, so it can be run with nix-shell.

I started doing this because I got sick of writing wrappers like the above, and even started running into programs that did not have a flag to override where they look for configuration.

seagreen commented:
I'm far from a NixOS expert, but I don't know if it's a good idea to put this into Nixpkgs `master` without testing it somewhere else first. As other people in this thread have mentioned there are already tools for managing dotfiles, and we can't know if NixUP will be better at this time.

Would it be possible to build this as a separate service or as a fork of `Nixpkgs`, and only merge it into master once we know it's the clear winner over the other options?

ixxie commented:
@ts468 - I will donate €50 for this cause, as soon as I figure out how to get PayPal to work xD

teto commented:
To sum up we have
- https://github.com/rycee/home-manager/
- https://github.com/sheenobu/nix-home
- nixup
I am looking for a solution where I could do   `programs.zsh.shellAliases = [ .. ]` and it would generate it into $ZDOTDIR/.zshrc for instance.
nix-home looks very basic (like a versioned repo) while home-manager and nixup look better suited. I am tempted to go for the home-manager way as it seems easier to contribute (it has own repo with issues/PRs).

ts468 committed:
nixos pam_env: use conffile instead of envfile

Change the internals of how the PAM module "pam_env" is configured.
So far pam_env read simple KEY=VALUE pairs from a file specified with the
option "envfile". Now, pam_env is switched to use the option "conffile" which
reads a file that allows more complex configurations. For example the file
read by "conffile" can contain special variables that are expanded by pam_env.

The effect of the switch to using "conffile" should be purely internal. On the
one hand the pam_env option "readenv=0" is used to prevent pam_env from loading
the default environment file at "/etc/environment" that would pam_env would
fall back to now. On the other hand the mechanism of creating the KEY=VALUE
pairs for the "envfile" option is adapted to fit the new format of the
"conffile" option. So there should only be a visible effect if a VALUE contained
any special variables like "@{HOME}" or "@{SHELL}" that are expanded now by
pam_env.

ts468 committed:
nixos virtualisation: make networking.useHostResolvConf overridable

ts468 committed:
nixos pam: add option "sessionCommands"

Add the option "sessionCommands" to security.pam.services. "sessionCommands"
gets executed by PAM right before the systemd user session is started.

ts468 committed:
nixos programs: clean-up "let ... in" abstractions

michalrus commented:
If anyone wants something _waaaaaaaaaaay_ easier, then:

* immutable user envs, declared in `configuration.nix`: https://github.com/michalrus/dotfiles/blob/bcc6d970c319568efc9ca25740d0d08f16217f6e/nixos-config/modules/user-profiles.nix

* very primitive dotfiles w/ “profiles” declared in `configuration.nix` (you can still use them easily on non-NixOS): https://github.com/michalrus/dotfiles/blob/bcc6d970c319568efc9ca25740d0d08f16217f6e/nixos-config/modules/dotfiles.nix

* sample usage: https://github.com/michalrus/dotfiles/blob/bcc6d970c319568efc9ca25740d0d08f16217f6e/nixos-config/machines/desktop/base-mine.nix#L156-L160

Cheers,

ts468 committed:
Add a type "relativePath".

ts468 committed:
Prepare nixos-rebuild.sh for NixUP

Nixos-rebuild.sh is generalized to be equally usable for NixOS and
for NixUP. Nixos-rebuild.sh is also renamed to nix__-rebuild.sh.
The changes do not affect the behaviour for NixOS.

ts468 committed:
Prepare switch-to-configuration.pl for NixUP

switch-to-configuration.sh is generalized to be equally usable
for NixOS and for NixUP.
The changes do not affect the behaviour for NixOS.

ts468 committed:
nixos nixup: add NixUP initialisation hooks

Add the configuration option "config.nixup.enable" which integrates the
activation scripts of NixUP into the user login and session handling of NixOS,
and provides the "nixup-rebuild" tool for managing NixUP.

ts468 committed:
Init NixUP.

Add the NixUP module system.

ts468 committed:
init: nixup-tools

ts468 committed:
nixup: add module: nixup-tools

ts468 commented:
A new and re-based version of NixUP arrived!

It would be great if you could review and test the changes!

Special attention should go into reviewing:
- nixos-rebuild.sh (verify to have no effect on NixOS)
- switch-to-configuration.pl (verify to have no effect on NixOS)
- resource-manager.c (security review as resource-manager runs with setuid=0)

Many thanks in advance!

vcunat commented:
Hydra for tests: https://hydra.nixos.org/eval/1410967?compare=1405848

davidak commented:
This looks amazing. I also donated 50€.

Any progress since last comment?

nbp commented:
@ts468 I see that I am being asked for review, but at the same time I see 2000 lines addition. Is there a way to split this work in multiple digest PR?

bobvanderlinden commented on nixos/modules/installer/tools/nix__-rebuild.sh:
```diff
@@ -19,73 +31,80 @@ fast=
         showSyntax
         ;;
-      switch|boot|test|build|dry-build|dry-run|dry-activate|build-vm|build-vm-with-bootloader)
+      switch_*|boot_nixos|login_nixup|test_nixos|build_*|dry-build_nixos|dry-run_*|dry-activate_nixos|build-vm_nixos|build-vm-with-bootloader_nixos|edit_nixos)
```
Doesn't this (and changes of commandline flags below) break the existing usage of `nixos-rebuild`?

throwup commented:
Sorry if I am asking something that should already be clear but is this:
> @bjornfor Thanks for your interest! Indeed, it is planed to make all NixUP options that are available under "~/.nixup/profile.nix" also available in the NixOS configuration, probably under "users.extraUsers..profile". However, I'm just not so sure how to merge the options defined in NixOS and NixUP, yet. Should they just be added up? Should the options defined in NixOS be defaults that can be overwritten with NixUP, or should that be prohibited? If you have any suggestions or ideas that would be great! :)

still a feature you intend to add? I like to define my system using a single git repository and would rather not link stuff to home directories manually. I read in another message, that this, based on how it's implemented, might "automatically" work, but I am not proficient enough at nix to judge that.

timokau commented:
What is blocking this right now? Is there any way to help push this forward?

dasJ commented:
Yes, it is possible this gets into 18.03? Apart from some conflicts I don't see anything keeping it from merged

IvanMalison commented:
@ts468 status?

drvink commented:
Is there a reason this has remained unmerged after nearly three years? The author's work is going to waste--it already needs *another* rebase and forward-porting effort now. It would be nice to at least hear from the Nix maintainers whether there are plans to integrate similar functionality.

Nadrieril commented:
I personally think home-manager would be a better basis for this, as it is more featureful and actively developped

IvanMalison commented:
> Is there a reason this has remained unmerged after nearly three years? The author's work is going to waste--it already needs another rebase and forward-porting effort now. It would be nice to at least hear from the Nix maintainers whether there are plans to integrate similar functionality.

> I personally think home-manager would be a better basis for this, as it is more featureful and actively developped

It would still be nice to hear something official from the maintainers, so that users of whichever system (be it nixup or homemanager) know which one is likely to be maintained and usable moving forward.

srghma commented:
such a good project have

sheyll commented:
Bumped into this, any progress?

tbenst commented:
If I understand correctly, one trade-off of NixUP vs home-manager is that NixUP requires NixOS whereas home-manager supports eg Ubuntu. Is my impression accurate? Hopefully, whatever solution is decided on will support other linux distros for home management, as some users will always have to deal with some non-NixOS systems.

8573 commented:
Are there any similar proposals for declarative user environment management (or simply declarative dotfile management) that would be configured on a system-wide basis, in `/etc/nixos/configuration.nix` (e.g., in the `users` option hierarchy), rather than on a per-user basis, in configuration files inside the users' managed home directories?

kalbasit commented:
@8573 home-manager is capable of configuring dotfiles from the NixOS configuration located at `/etc/nixos/configuration.nix`. See the home-manager module [here](https://github.com/rycee/home-manager/blob/master/nixos/default.nix)

Personally, I took it a step further. I made a [home module](https://github.com/kalbasit/system/tree/master/modules/home) (similar to NixOS modules) controlled by options under the namespace `mine`. I have another NixOS module that sets the `home-manager.users.<name>` to a function call that returns the set of options controlling the home module mentioned above. See this [host](https://github.com/kalbasit/system/blob/6e2e56c0b28d7918ff4ae4d80352b75f7c728045/hosts/hades/home.nix) as an example and I set the users [here](https://github.com/kalbasit/system/blob/6e2e56c0b28d7918ff4ae4d80352b75f7c728045/modules/nixos/general/users.nix#L84).

 I set the options per-host in the home.nix files [here]() 


To give credit where credit is due, my system is inspired and based on @dustinlacewell's [dotfiles](https://github.com/dustinlacewell/dotfiles) and the IRC community.

mudriion commented:
Nice to see one official user env/file management in Nixos
Any progress?

mmahut commented:
Looks like @ts468 is not active for some time. Given the lack of larger consensus on this pull request, maybe we should close it and move on?

Infinisil commented:
Sounds good, if somebody wants to pick up on the authors work they're still welcome to do so, probably best in a new PR. Note however that [home-manager](https://github.com/rycee/home-manager) is doing a very good job at this already, in a way that doesn't need integration in nixpkgs.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/why-isnt-more-of-home-manager-merged-into-nixpkgs/6096/3
